<launch>
  <arg name="rviz" default="true"/>
  <param name="/use_sim_time" value="true"/>

  <!-- Run optic flow -->
  <node name="optic_flow_node" pkg="optic_flow_node" type="wide_field_optic_flow_node" output="screen" >
      <param name="image_center_x" value="320"/>
      <param name="image_center_y" value="240"/>
      <param name="inner_ring_radius" value="235"/>
      <param name="ring_dr" value="10"/>
      <param name="num_ring_points" value="160"/>
      <param name="num_rings" value="5"/>
      <param name="pyr_window_size" value="50"/>
      <param name="pixel_scale" value="215"/>
      <param name="alpha" value="0.7"/>
      <!-- Blur size must be an int  -->
      <param name="if_blur" value="true"/>
      <param name="blur_size" value="5"/>
  </node>

  <!-- Run radar, PCL filter and convert to laserscan -->
  <include file="$(find pointcloud_to_laserscan)/launch/pcl2laserscan.launch" />

  <node name="kf_nearness_node" pkg="nearness_estimation" type="kf_nearness_node" output="screen" >
      <param name="covariance_process" value="100"/>
      <param name="covariance_optic_flow" value="5"/>
      <param name="covariance_radar" value="0.001"/>
      <param name="threshold" value="0.3"/>
      <param name="minimum_velocity" value="0.4"/>
      <param name="lowess_smoothness" value="0.2"/>
  </node>


  <!-- Convert estimated laser nearness to PCL and run octomap server-->
  <include file="$(find pointcloud_to_laserscan)/launch/laserscan2pcl.launch" />

  <!-- Odometry and transforms-->
  <include file="$(find robot_localization)/launch/jackal_localization.launch" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_imu" args="0.0 0.0 0.0 0.0 0.0 0 1.0 base_link imu_link" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_radar" args="0.27 0.0 0.0 0.0 0.0 0 1.0 base_link base_radar_link" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_laser" args="0.1 0.0 0.0 0.0 0.0 1.0 0.0 base_link laser" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_oflow" args="-0.17 0.0 0.0 0 0 0 1 base_radar_link oflow_frame" />
  <!-- Need to rotate oflow frame to display estimated depth in laser msg -->
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_oflow_laser" args="0.0 0.0 0.0 1 0 0 0 oflow_frame oflow_laser_frame" />
  <node pkg="tf2_ros" type="static_transform_publisher" name="tf_map" args="0.0 0.0 0.0 0.0 0.0 0 1.0 map odom" />

  <!-- Rviz -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find pointcloud_to_laserscan)/rviz/visualize_octomap.rviz"/>

</launch>
